### PR TITLE
Add Used() method with delegate to jira client

### PR DESCRIPTION
**Pull request by jmguzik**
https://github.com/kubernetes/test-infra/pull/25669 follow up.

/cc @stevekuznetsov @petr-muller @alvaroaleman 

Signed-off-by: Jakub Guzik <jguzik@redhat.com>

### CODE REVIEW

**Comment by jmguzik** [comment]
@stevekuznetsov by the way I spent some time analyzing the rest of the clients. I found two that can be interesting, but personally I think there is no need to modify them:

- kubernetes client - I think can be difficult to modify and it's used in one plugin. The call is somewhere deep below calls to other clients (eg Github client) so adding `Used()` does not bring much benefit.
- slack client - similar situation as above, maybe it's easier to modify it but there is a use of other clients multiple times before that call.

If we ignore adding `Used()` to these clients everything will work as we added them. I also analyzed the need for modification of the return type. I think we can also ignore it as some external calls are done only in minor plugins (like `cat` or `dog`).

I propose the next step (after merging this PR) would be to actually enable metric labels.

File: prow/jira/jira.go [hunk 1]
```diff
@@ -107,7 +110,7 @@ func NewClient(endpoint string, opts ...Option) (Client, error) {
 	}
 
 	jiraClient, err := jira.NewClient(retryingClient.StandardClient(), endpoint)
-	return &client{upstream: jiraClient, url: endpoint}, err
+	return &client{delegate: &delegate{upstream: jiraClient, url: endpoint}, logger: log}, err
```
Comment (alvaroaleman) [inline prow/jira/jira.go:113]: It would be nice to find a way to set this without having to implement it on every single method, because that is almost guaranteed to rot over time and be incomplete.

How about:
* We add an `upstreamConstructor func(t http.RoundTripper) *jiraClient`
* In `WithFields` etc we call it with a roundTripper that sets a `used` variable upon first use
* `Used` on the client uses said variable

Comment (jmguzik) [reply prow/jira/jira.go:113]: Nice idea, thanks. I'll rework this.

Comment (jmguzik) [reply prow/jira/jira.go:113]: @alvaroaleman After analysis I can't see this working with jira client extracted to the `delegate` struct. I want to have a common client (for all plugins), but a unique `used` field. Maybe I missed something?

Comment (alvaroaleman) [reply prow/jira/jira.go:113]: That is why I suggest the `upstreamConstructor`

Comment (jmguzik) [reply prow/jira/jira.go:113]: But that will create a new jira client every time called, right?

Comment (alvaroaleman) [reply prow/jira/jira.go:113]: Yes, it doesn't allow to update the RoundTripper

Comment (jmguzik) [reply prow/jira/jira.go:113]: @alvaroaleman got back to this PR to make discussed change, almost done but I see one potential problem:
https://github.com/kubernetes/test-infra/blob/34e7342a4b709f2b449f2119a049cb37865bad23/prow/jira/jira.go#L109
returns not only the client but also an error. That would require `ForPlugin` to return an error as well, and possibly also `NewAgent` in https://github.com/kubernetes/test-infra/blob/34e7342a4b709f2b449f2119a049cb37865bad23/prow/plugins/plugins.go#L204
which will propagate to hook...
I think it is a pretty serious change in logic, and before I continue I would like to have your opinion.

File: prow/jira/jira.go [hunk 2]
```diff
@@ -121,9 +134,27 @@ func (u userAgentSettingTransport) RoundTrip(r *http.Request) (*http.Response, e
...
+	c.used = true
+	return c.upstream.RoundTrip(r)
+}
+
 type client struct {
-	url      string
-	upstream *jira.Client
+	logger    *logrus.Entry
+	upstream  *jira.Client
+	transport *clientUsedTransport
```
Comment (stevekuznetsov) [inline prow/jira/jira.go:150]: nit: can this be an interface - I don't think we necessarily want to expose the delegate or transporting function in this field

```go
type used interface {
    Used() bool
}
```

**Review by stevekuznetsov — COMMENTED** [review]
Looks reasonable - just one comment!

File: prow/jira/jira.go [hunk 3]
```diff
@@ -121,9 +134,27 @@ func (u userAgentSettingTransport) RoundTrip(r *http.Request) (*http.Response, e
 	return u.upstream.RoundTrip(r)
 }
 
+type clientUsedTransport struct {
+	used     bool
+	upstream http.RoundTripper
+}
+
+func (c clientUsedTransport) RoundTrip(r *http.Request) (*http.Response, error) {
```
Comment (alvaroaleman) [inline prow/jira/jira.go:142]: This needs to be threadsafe - It happens to not matter today since there is only one plugin that uses this and we reconstruct per event, but that might be different in the future

File: prow/jira/jira.go [hunk 4]
```diff
@@ -214,11 +245,47 @@ func (jc *client) JiraURL() string {
...
+func (jc *client) ForPlugin(plugin string) Client {
+	pluginLogger := jc.logger.WithField("plugin", plugin)
+	retryingClient := retryablehttp.NewClient()
+	usedFlagTransport := &clientUsedTransport{
+		upstream: retryingClient.HTTPClient.Transport,
+	}
+	retryingClient.HTTPClient.Transport = usedFlagTransport
+	retryingClient.Logger = &retryableHTTPLogrusWrapper{log: pluginLogger}
+	// ignore error as url.String() was passed to the delegate
+	jiraClient, _ := newJiraClient(jc.url, jc.options, retryingClient)
```
Comment (alvaroaleman) [inline prow/jira/jira.go:280]: Can we log the error here? We don't expect it to be possible that this errors, but for example if the jira lib gets upgraded and then errors for whatever other reason, this could happen and would be a PITA to debug

File: prow/jira/jira.go [hunk 5]
```diff
@@ -107,8 +99,29 @@ func NewClient(endpoint string, opts ...Option) (Client, error) {
...
+	}
+	retryingClient := retryablehttp.NewClient()
+	usedFlagTransport := &clientUsedTransport{
+		upstream: retryingClient.HTTPClient.Transport,
+	}
+	retryingClient.HTTPClient.Transport = usedFlagTransport
+	retryingClient.Logger = &retryableHTTPLogrusWrapper{log: log}
+
+	jiraClient, err := newJiraClient(endpoint, o, retryingClient)
+	url := jiraClient.GetBaseURL()
```
Comment (alvaroaleman) [inline prow/jira/jira.go:123]: Please return here if err != nil, otherwise this might end up being a nil pointer dereference

File: prow/jira/jira.go (hunk 2 above)
Comment (jmguzik) [reply prow/jira/jira.go:150]: Reasonable but segfaulting :) Thanks

File: prow/jira/jira.go (hunk 3 above)
Comment (jmguzik) [reply prow/jira/jira.go:142]: Sorry, I don't get it. Why threadsafe? Every time ForPlugin is called, a completely new transport along with Jira client is created. Maybe missed something...

File: prow/jira/jira.go [hunk 6]
```diff
@@ -214,11 +263,49 @@ func (jc *client) JiraURL() string {
...
+		clientUsed: jc.clientUsed,
+		upstream:   jc.upstream,
+		logger:     jc.logger.WithFields(fields),
+		delegate:   jc.delegate,
+	}
+}
+
+// ForPlugin clones the client, keeping the underlying delegate the same but adding
+// a plugin identifier and log field
+func (jc *client) ForPlugin(plugin string) Client {
```
Comment (jmguzik) [inline prow/jira/jira.go:289]: There is a bug somewhere in that function or something is not initialized for integration tests as they are failing on nil. I'll try to investigate later on.

Comment (jmguzik) [reply prow/jira/jira.go:289]: Discovered that jiraClient can be nil in Agent. Added if that should solve the problem.

File: prow/jira/jira.go [hunk 7]
```diff
@@ -121,9 +139,40 @@ func (u userAgentSettingTransport) RoundTrip(r *http.Request) (*http.Response, e
...
 
+type clientUsedTransport struct {
+	used     bool
+	m        sync.Mutex
+	upstream http.RoundTripper
+}
+
+func (c *clientUsedTransport) RoundTrip(r *http.Request) (*http.Response, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
```
Comment (stevekuznetsov) [inline prow/jira/jira.go:150]: By using `defer` here we have just made a single-threaded client :)

Also as a nit, even if the `defer` was not containing the `c.upstream` call, it's about 5-10x slower than explicitly unlocking in cases like this. I think we should:

```go
c.m.Lock()
c.used = true
c.m.Unlock()
c.upstream.RoundTrip(r)
```

File: prow/jira/jira.go [hunk 8]
```diff
@@ -214,11 +263,49 @@ func (jc *client) JiraURL() string {
...
+	retryingClient := retryablehttp.NewClient()
+	usedFlagTransport := &clientUsedTransport{
+		m:        sync.Mutex{},
+		upstream: retryingClient.HTTPClient.Transport,
+	}
+	retryingClient.HTTPClient.Transport = usedFlagTransport
+	retryingClient.Logger = &retryableHTTPLogrusWrapper{log: pluginLogger}
+	// ignore error as url.String() was passed to the delegate
+	jiraClient, err := newJiraClient(jc.url, jc.options, retryingClient)
+	pluginLogger.Error(err)
```
Comment (stevekuznetsov) [inline prow/jira/jira.go:300]: nit: this will put the error in a field called `msg` which people might miss, as `pluginLogger.WithError(err).Error("invalid Jira URL")` will put it in a field `err`

File: prow/jira/jira.go (hunk 7 above)
Comment (jmguzik) [reply prow/jira/jira.go:150]: Yes, you are right. I wanted to change this but then forgot. Thanks.

Comment (alvaroaleman) [reply prow/jira/jira.go]: Agree re the single threading point. The performance issues `defer` had were mostly fixed some time back btw: https://rakyll.org/inlined-defers/

File: prow/jira/jira.go [hunk 9]
```diff
@@ -107,8 +100,33 @@ func NewClient(endpoint string, opts ...Option) (Client, error) {
...
+		m:        sync.Mutex{},
+		upstream: retryingClient.HTTPClient.Transport,
+	}
+	retryingClient.HTTPClient.Transport = usedFlagTransport
+	retryingClient.Logger = &retryableHTTPLogrusWrapper{log: log}
+
+	var url url.URL
+	jiraClient, err := newJiraClient(endpoint, o, retryingClient)
+	if err != nil {
+		url = jiraClient.GetBaseURL()
```
Comment (alvaroaleman) [inline prow/jira/jira.go:127]: We can not access a non-err return if we had an error, this needs to be an `url.Parse(endpoint)` or some such and also consider the possibility that endpoint might not be a valid url

**Comment by jmguzik** [comment]
Of course, it failed 
/retest

File: prow/jira/jira.go [hunk 10]
```diff
@@ -107,8 +100,33 @@ func NewClient(endpoint string, opts ...Option) (Client, error) {
...
+	}
+	return &client{delegate: &delegate{url: url.String(), options: o}, logger: log, upstream: jiraClient, clientUsed: usedFlagTransport}, err
```
[the remark this answers is not in the document]
Comment (jmguzik) [reply prow/jira/jira.go:129]: We do like this because it was done like this in the past. I just preserved the behavior. I have no problem with changing it to `return nil, err`

File: prow/jira/jira.go (hunk 9 above)
Comment (jmguzik) [reply prow/jira/jira.go:127]: This point I don't get. Can you rephrase?

Comment (alvaroaleman) [reply prow/jira/jira.go:127]: if `newJiraClient` returns a non-nil error, you can now access `jiraClient`, because whenever you get an `err` from a function, you can not make any assumption about the other returns of that function. Here, we assume that even in the error case it returns a non-nil `jiraClient`

File: prow/jira/jira.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (alvaroaleman) [reply prow/jira/jira.go:129]: True, but I think that was just for brevity and happened to be fine, because we didn't have to access the `jiraClient` in the error case

**Review by alvaroaleman — APPROVED** [review]
Thanks for your work!

**Comment by jmguzik** [comment]
Thanks, @alvaroaleman !
Unfortunately, there was a conflict. I had to rebase.

**Review by alvaroaleman — APPROVED** [review]

**Comment by alvaroaleman** [comment]
/lgtm cancel
The fakeclient doesn't fulfill the interface anymore :upside_down_face:

**Comment by jmguzik** [comment]
I actually used that fake client :)
